src/Trie.mo

```diff
@@ -25,58 +17,71 @@ import List "List";
-it can actually construct valid leaves.  Ongce distinguished, subsets
-of keys tend to remain distinguished by the presence of these branches.
-
-### Cached sizes
-
-At each branch and leaf, we use a stored size to support a
-memory-efficient `toArray` function, which itself relies on
-per-element projection via `nth`; in turn, `nth` directly uses the
-O(1)-time function `size` for achieving an acceptable level of
-algorithmic efficiently.  Notably, leaves are generally lists of
-key-value pairs, and we do not store a size for each Cons cell in the
-list.
 
+///
+/// ## Representation
+///
+///
+/// A hash trie is a binary trie, where each (internal) branch node
+/// represents having distinguished its key-value pairs on a single bit of
+/// the keys.
+///
+/// By following paths in the trie, we determine an increasingly smaller
+/// and smaller subset of the keys.
+///
+/// Each leaf node consists of an association list of key-value pairs.
+///
+/// We say that a leaf is valid if it contains no more than MAX_LEAF_SIZE
+/// key-value pairs.
+///
+/// Each non-empty trie node stores a size; we discuss that more below.
+///
+/// ### Adaptive depth
+///
+/// For small mappings, the trie structure consists of a single
+/// leaf, which contains up to MAX_LEAF_SIZE key-value pairs.
+///
+/// By construction, the algorithms enforce an invariant that no
+/// leaf ever contains more than MAX_LEAF_SIZE key-value pairs: the
+/// function `leaf` accepts a list, but subdivides it with branches until
+/// it can actually construct valid leaves.  Ongce distinguished, subsets
+/// of keys tend to remain distinguished by the presence of these branches.
+///
+/// ### Cached sizes
+///
+/// At each branch and leaf, we use a stored size to support a
+/// memory-efficient `toArray` function, which itself relies on
+/// per-element projection via `nth`; in turn, `nth` directly uses the
+/// O(1)-time function `size` for achieving an acceptable level of
+/// algorithmic efficiently.  Notably, leaves are generally lists of
+/// key-value pairs, and we do not store a size for each Cons cell in the
+/// list.
+///
 
-### Details
+let MAX_LEAF_SIZE = 8; // to do -- further profiling and tuning
 
-Below, we define the types used in the representation:
+/// binary hash tries: either empty, a leaf node, or a branch node
+public type Trie<K,V> = {
+  #empty  ;
+  #leaf   : Leaf<K,V> ;
+  #branch : Branch<K,V> ;
+};
 
-*/
+/// leaf nodes of trie consist of key-value pairs as a list.
+public type Leaf<K,V> = {
+  size   : Nat ;
+  keyvals : AssocList<Key<K>,V> ;
+};
 
-//let MAX_LEAF_SIZE = 4; // worse than 8, slightly
-public let MAX_LEAF_SIZE = 8; // <-- beats both 4 and 16 for me, now
-//let MAX_LEAF_SIZE = 16;
-//let MAX_LEAF_SIZE = 32;
+/// branch nodes of the trie discriminate on a bit position of the keys' hashes.
+/// we never store this bitpos; rather,
+/// we enforce a style where this position is always known from context.
```

Comment:
Is this a comment that is relevant to a user of this library? For implementation notes that shouldn't matter for a user we should continue to use regular (`//`) comments.

Reply:
Well, the (public) type of the trie is exposed, and the type has invariants.  These comments explain those invariants.

OTOH, if you just use the methods and never inspect the structure, these invariants are just maintained for the user, and these notes are not relevant.

It seems like a grey area to me.

Reply:
Keep in mind that this module is distinct from `TrieMap`, which gives the more friendly (and opaque) interface, where these invariants need not be mentioned at all.  I view this module here as a more "low level" one, where the user has the option to depend on these invariants, call the method that checks for them being true, etc.

Reply:
Or, perhaps there's a compromise where some of this discussion can move into an inner public module, and its documentation (e.g., `public module CheckInvariants { ... }`, but a better name)?  Would require nested module documentation support in the `mo-doc` tooling, but that seems like it will come at some point soon enough.

Reply:
If this module is basically an implementation detail for the user facing TrieMap type I don't mind the comment here.